nixos/modules/security/pass.nix

```diff
@@ -0,0 +1,200 @@
+          '';
+        };
+        mode = lib.mkOption {
+          type = types.str;
+          default = "400";
+          description = ''
+            Permission mode of the secret <literal>path</literal>.
+          '';
+        };
+        user = lib.mkOption {
+          type = types.str;
+          default = "root";
+          description = ''
+            Owner of the secret <literal>path</literal>.
+          '';
+        };
+        group = lib.mkOption {
+          type = types.str;
+          default = "root";
+          description = ''
+            Group of the secret <literal>path</literal>.
+          '';
+        };
+        pipe = lib.mkOption {
+          type = types.nullOr types.str;
+          default = null;
+          description = ''
+            Shell command taking the deciphered secret on its standard input
+            and which must put on its standard output
+            the actual material to be installed.
+            This allows to decorate the secret with non-secret bits.
+          '';
+        };
+        path = lib.mkOption {
+          type = types.str;
+          default = name;
+          apply = p: if match "^/.*" p == null then "/run/pass-secrets/"+p+"/file" else p;
+          description = ''
+            The path on the target system where the secret is installed to.
+            Any non-absolute path is relative to <filename>/run/pass-secrets</filename>.
+            Default to the name of the secret.
+          '';
+        };
+        service = lib.mkOption {
+          type = types.str;
+          default = "secret-" + escapeUnitName name + ".service";
+          description = ''
+            The name of the systemd service.
+            Useful to put constraints like <literal>after</literal> or <literal>wants</wants>
+            into services requiring this secret.
+          '';
+        };
+        postStart = lib.mkOption {
+          type = types.lines;
+          default = "";
+          example = "systemctl reload nginx.service";
+          description = ''
+            Commands to run after new secrets go live.
+            Typically the web server and other servers using secrets
+            need to be reloaded.
+          '';
+        };
+        postStop = lib.mkOption {
+          type = types.lines;
+          default = "";
+          example = ''shred -u "$secret_file"'';
+          description = ''
+            Commands to run after stopping the service.
+            Typically removing a persistent secret.
+          '';
+        };
+      };
+    }));
+  };
+};
+config = lib.mkIf (pass.secrets != {}) {
+  systemd.services =
+    lib.mapAttrs' (target: secret:
+      lib.nameValuePair (lib.removeSuffix ".service" secret.service) {
+        description = "Install secret ${secret.path}";
+        script = ''
+          set -o pipefail
+          set -eux
+          decrypt() {
+            {
+            ${pkgs.gnupg}/bin/gpg --batch --pinentry-mode loopback \
```

Comment:
Does this not create a gpg daemon? Should be stopped afterwards or do we do not bother? Or does systemd kills it when the service finishes?

Reply:
Oh! indeed... and this gpg-agent would be destroyed by `pam_systemd` removing `/run/user/0/` when the last `root` logouts. This has given me the idea to leverage `gpg-agent` to avoid writting the passphrase on disk (though it's still possible if one wants a persistent setup). For that I had to rewrite most of the module, and also decided to rename it `security.gnupg` which is a clearer name wrt. what it does now. It was a bit hard to find a clean way to bypass the wipe out of `pam_systemd` to preserve `gpg-agent` but I may have found one, as explained in the comments. Thanks for bringing this to my attention.